routing/http/types/address.go

```diff
@@ -0,0 +1,319 @@
+func (a *Address) HasProtocol(proto string) bool {
+	proto = strings.ToLower(proto)
+
+	if a.url != nil {
+		scheme := strings.ToLower(a.url.Scheme)
+
+		switch proto {
+		case "http":
+			// "http" matches both http and https URLs
+			return scheme == "http" || scheme == "https"
+		case "https":
+			return scheme == "https"
+		case "tls":
+			// TLS matches https URLs
+			return scheme == "https"
+		default:
+			return scheme == proto
+		}
+	} else if a.multiaddr != nil {
+		protocols := a.Protocols()
+
+		switch proto {
+		case "http":
+			// "http" matches /http, including /tls/http
+			for _, p := range protocols {
+				if p == "http" {
+					return true
+				}
+			}
+		case "https":
+			// "https" matches /https or the combination /tls/http
+			hasTLS := false
+			hasHTTP := false
+			for _, p := range protocols {
+				if p == "https" {
+					return true
+				}
+				if p == "tls" {
+					hasTLS = true
+				}
+				if p == "http" {
+					hasHTTP = true
+				}
+			}
+			return hasTLS && hasHTTP
+		case "tls":
+			// "tls" matches any multiaddr with /tls
+			for _, p := range protocols {
+				if p == "tls" {
+					return true
+				}
+			}
+		default:
+			for _, p := range protocols {
+				if p == proto {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+// ToMultiaddr attempts to convert an HTTP(S) URL to a multiaddr for backward compatibility.
+// Returns nil if the address cannot be converted (e.g., non-HTTP schemes or invalid addresses).
+// This is a temporary compatibility layer for the transition period while existing software
+// expects multiaddrs with /http protocol to signal HTTP retrieval support.
+func (a *Address) ToMultiaddr() multiaddr.Multiaddr {
+	// If already a multiaddr, return it as-is
+	if a.IsMultiaddr() {
+		return a.multiaddr
+	}
+
+	// If not a URL, cannot convert
+	if !a.IsURL() || a.url == nil {
+		return nil
+	}
+
+	// Only convert http/https URLs
+	scheme := strings.ToLower(a.url.Scheme)
+	if scheme != "http" && scheme != "https" {
+		return nil
+	}
+
+	// Parse hostname and port
+	host := a.url.Hostname()
+	port := a.url.Port()
+
+	// Set default ports if not specified
+	if port == "" {
+		if scheme == "https" {
+			port = "443"
+		} else {
+			port = "80"
+		}
+	}
+
+	// Determine address type
+	var addrProto string
+	if ip := net.ParseIP(host); ip != nil {
+		// Use IP-specific protocols for IP addresses
+		if ip.To4() != nil {
+			addrProto = "ip4"
+		} else {
+			addrProto = "ip6"
+		}
+	} else {
+		// Use generic /dns for domain names (resolves to both IPv4 and IPv6)
+		addrProto = "dns"
+	}
+
+	// Build multiaddr string
+	var maStr string
+	if scheme == "https" {
+		// For HTTPS, use /https as this is what existing HTTP providers
+		// announce on IPNI, so we follow same convention for backward-compatibility
+		maStr = fmt.Sprintf("/%s/%s/tcp/%s/https", addrProto, host, port)
+	} else {
+		// For HTTP, use /http
+		maStr = fmt.Sprintf("/%s/%s/tcp/%s/http", addrProto, host, port)
+	}
```

Comment:
Since scheme can only be "http" or "https" just use scheme, instead of the if-else.
```suggestion
	// For HTTPS, use /https as this is what existing HTTP providers
	// announce on IPNI, so we follow same convention for backward-compatibility
	maStr := fmt.Sprintf("/%s/%s/tcp/%s/%s", addrProto, host, port, scheme)
```